collector/mdadm.go

```diff
@@ -0,0 +1,263 @@
+// +build !nomdadm
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"regexp"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/log"
+)
+
+var (
+	statusfile = "/proc/mdstat"
+	statuslineRE = regexp.MustCompile(`(\d+) blocks .*\[(\d+)/(\d+)\] \[[U_]+\]`)
+	buildlineRE = regexp.MustCompile(`\((\d+)/\d+\)`)
+)
+
+type mdstatus struct {
+	mdName       string
+	isActive     bool
+	disksActive  int64
+	disksTotal   int64
+	blocksTotal  int64
+	blocksSynced int64
+}
+
+type mdadmCollector struct {}
+
+func init() {
+	Factories["mdadm"] = NewMdadmCollector
+}
+
+func evalStatusline(statusline string) (active, total, size int64, err error) {
+
+	matches := statuslineRE.FindStringSubmatch(statusline)
+
+	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
+	if len(matches) < 3+1 {
+		return 0, 0, 0, fmt.Errorf("too few matches found in statusline: %s", statusline)
+	} else {
+		if len(matches) > 3+1 {
+			return 0, 0, 0, fmt.Errorf("too many matches found in statusline: %s", statusline)
+		}
+	}
+
+	size, err = strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+
+	total, err = strconv.ParseInt(matches[2], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+	active, err = strconv.ParseInt(matches[3], 10, 64)
+	if err != nil {
+		return 0, 0, 0, fmt.Errorf("%s in statusline: %s", err, statusline)
+	}
+
+	return active, total, size, nil
+}
+
+// Gets the size that has already been synced out of the sync-line.
+func evalBuildline(buildline string) (int64, error) {
+	matches := buildlineRE.FindStringSubmatch(buildline)
+
+	// +1 to make it more obvious that the whole string containing the info is also returned as matches[0].
+	if len(matches) < 1+1 {
+		return 0, fmt.Errorf("too few matches found in buildline: %s", buildline)
+	} else {
+		if len(matches) > 1+1 {
+			return 0, fmt.Errorf("too many matches found in buildline: %s", buildline)
+		}
+	}
+
+	syncedSize, err := strconv.ParseInt(matches[1], 10, 64)
+
+	if err != nil {
+		return 0, fmt.Errorf("%s in buildline: %s", err, buildline)
+	} else {
+		return syncedSize, nil
+	}
+}
+
+// Parses an mdstat-file and returns a struct with the relevant infos.
+func parseMdstat(mdstatusfilepath string) ([]mdstatus, error) {
+	content, err := ioutil.ReadFile(mdstatusfilepath)
+	if err != nil {
+		return []mdstatus{}, fmt.Errorf("error parsing %s: %s", statusfile, err)
+	}
+
+	mdStatusFile := string(content)
```

Comment:
The comment I made here also applies to other variables, like `mdstatusfilepath` (`mdStatusFilePath`) above, or `mdstates` (`mdStates`). Camel case on word boundaries please.